File: processor/src/main/java/com/qualityminds/lazyval/processor/internal/codegen/beanvalidation/StringValidatorBuilder.java
package com.qualityminds.lazyval.processor.internal.codegen.beanvalidation;

import com.palantir.javapoet.*;
import com.qualityminds.lazyval.processor.internal.codegen.GeneratedStamp;
import com.qualityminds.lazyval.processor.spi.GeneratorResult;
import com.qualityminds.lazyval.processor.spi.ValidatedGeneratorElement;

import javax.lang.model.element.Modifier;
import javax.lang.model.type.TypeMirror;
import java.util.Set;
import java.util.stream.Stream;

class StringValidatorBuilder {

    private static final Set<String> STRING_TYPES = Set.of("String");

    private static final ClassName CONSTRAINT_VALIDATOR = ClassName.get("jakarta.validation", "ConstraintValidator");
    private static final ClassName CONSTRAINT_VALIDATOR_CONTEXT = ClassName.get("jakarta.validation", "ConstraintValidatorContext");

    static boolean supports(String typeName) {
        return STRING_TYPES.contains(typeName);
    }

    static Stream<GeneratorResult> build(ValidatedGeneratorElement element, String packageName) {
        return Stream.concat(
                buildPatternValidator(element, packageName),
                buildEmailValidator(element, packageName));
    }

    private static Stream<GeneratorResult> buildPatternValidator(ValidatedGeneratorElement element, String packageName) {
        TypeMirror lazyvalType = element.element().asType();
        String className = element.typeName().name() + "PatternValidator";

        ClassName patternAnnotation = ClassName.get("jakarta.validation.constraints", "Pattern");

        TypeSpec validator = TypeSpec.classBuilder(className)
                .addAnnotation(GeneratedStamp.forGenerator(BeanValidationGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(ParameterizedTypeName.get(
                        CONSTRAINT_VALIDATOR, patternAnnotation, TypeName.get(lazyvalType)))
                .addField(FieldSpec.builder(String.class, "regex", Modifier.PRIVATE).build())
                .addMethod(buildInitialize(patternAnnotation, "this.regex = constraintAnnotation.regexp()"))
                .addMethod(buildIsValid(element, lazyvalType, "regex"))
                .build();

        JavaFile javaFile = JavaFile.builder(packageName, validator).build();
        return ResultHelper.toResultStream(javaFile, packageName, className);
    }

    private static Stream<GeneratorResult> buildEmailValidator(ValidatedGeneratorElement element, String packageName) {
        TypeMirror lazyvalType = element.element().asType();
        String className = element.typeName().name() + "EmailValidator";

        ClassName emailAnnotation = ClassName.get("jakarta.validation.constraints", "Email");

        TypeSpec validator = TypeSpec.classBuilder(className)
                .addAnnotation(GeneratedStamp.forGenerator(BeanValidationGenerator.class))
                .addModifiers(Modifier.PUBLIC)
                .addSuperinterface(ParameterizedTypeName.get(
                        CONSTRAINT_VALIDATOR, emailAnnotation, TypeName.get(lazyvalType)))
                .addField(FieldSpec.builder(String.class, "regexp", Modifier.PRIVATE).build())
                .addMethod(buildInitialize(emailAnnotation, "this.regexp = constraintAnnotation.regexp()"))
                .addMethod(buildIsValid(element, lazyvalType, "regexp"))
                .build();

        JavaFile javaFile = JavaFile.builder(packageName, validator).build();
        return ResultHelper.toResultStream(javaFile, packageName, className);
    }

    private static MethodSpec buildInitialize(ClassName annotationType, String initStatement) {
        return MethodSpec.methodBuilder("initialize")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .addParameter(annotationType, "constraintAnnotation")
                .addStatement(initStatement)
                .build();
    }

    private static MethodSpec buildIsValid(ValidatedGeneratorElement element, TypeMirror lazyvalType, String fieldName) {
        return MethodSpec.methodBuilder("isValid")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(boolean.class)
                .addParameter(TypeName.get(lazyvalType), "value")
                .addParameter(CONSTRAINT_VALIDATOR_CONTEXT, "context")
                .beginControlFlow("if (value == null)")
                .addStatement("return true")
                .endControlFlow()
                .addStatement("return value.$L.matches($L)", element.accessor(), fieldName)
                .build();
    }
}

File: ksp/src/main/kotlin/com/qualityminds/lazyval/ksp/internal/codegen/beanvalidation/StringValidatorBuilder.kt
package com.qualityminds.lazyval.ksp.internal.codegen.beanvalidation

import com.qualityminds.lazyval.ksp.spi.GeneratorResult
import com.qualityminds.lazyval.ksp.spi.ValidatedKspGeneratorElement
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import java.util.stream.Stream

internal object StringValidatorBuilder {

    private val STRING_TYPES = setOf("String")

    private val CONSTRAINT_VALIDATOR = ClassName("jakarta.validation", "ConstraintValidator")
    private val CONSTRAINT_VALIDATOR_CONTEXT = ClassName("jakarta.validation", "ConstraintValidatorContext")

    fun supports(typeName: String): Boolean = typeName in STRING_TYPES

    fun build(element: ValidatedKspGeneratorElement, packageName: String): Stream<GeneratorResult> {
        return Stream.concat(
            buildPatternValidator(element, packageName),
            buildEmailValidator(element, packageName))
    }

    private fun buildPatternValidator(element: ValidatedKspGeneratorElement, packageName: String): Stream<GeneratorResult> {
        val lazyvalTypeName = element.element.toClassName()
        val className = "${element.typeName.name}PatternValidator"
        val patternAnnotation = ClassName("jakarta.validation.constraints", "Pattern")

        val typeSpec = TypeSpec.classBuilder(className)
            .addSuperinterface(
                CONSTRAINT_VALIDATOR.parameterizedBy(patternAnnotation, lazyvalTypeName)
            )
            .addProperty(PropertySpec.builder("regex", String::class, KModifier.PRIVATE)
                .mutable(true)
                .initializer("%S", "")
                .build())
            .addFunction(buildInitialize(patternAnnotation, "this.regex = constraintAnnotation.regexp"))
            .addFunction(buildIsValid(element, lazyvalTypeName, "regex"))
            .build()

        val fileSpec = FileSpec.builder(packageName, className).addType(typeSpec).build()
        return ResultHelper.toResultStream(fileSpec, packageName, className)
    }

    private fun buildEmailValidator(element: ValidatedKspGeneratorElement, packageName: String): Stream<GeneratorResult> {
        val lazyvalTypeName = element.element.toClassName()
        val className = "${element.typeName.name}EmailValidator"
        val emailAnnotation = ClassName("jakarta.validation.constraints", "Email")

        val typeSpec = TypeSpec.classBuilder(className)
            .addSuperinterface(
                CONSTRAINT_VALIDATOR.parameterizedBy(emailAnnotation, lazyvalTypeName)
            )
            .addProperty(PropertySpec.builder("regexp", String::class, KModifier.PRIVATE)
                .mutable(true)
                .initializer("%S", "")
                .build())
            .addFunction(buildInitialize(emailAnnotation, "this.regexp = constraintAnnotation.regexp"))
            .addFunction(buildIsValid(element, lazyvalTypeName, "regexp"))
            .build()

        val fileSpec = FileSpec.builder(packageName, className).addType(typeSpec).build()
        return ResultHelper.toResultStream(fileSpec, packageName, className)
    }

    private fun buildInitialize(annotationType: ClassName, initStatement: String): FunSpec {
        return FunSpec.builder("initialize")
            .addModifiers(KModifier.OVERRIDE)
            .addParameter("constraintAnnotation", annotationType)
            .addStatement(initStatement)
            .build()
    }

    private fun buildIsValid(element: ValidatedKspGeneratorElement, lazyvalTypeName: ClassName, fieldName: String): FunSpec {
        return FunSpec.builder("isValid")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Boolean::class)
            .addParameter("value", lazyvalTypeName.copy(nullable = true))
            .addParameter("context", CONSTRAINT_VALIDATOR_CONTEXT)
            .beginControlFlow("if (value == null)")
            .addStatement("return true")
            .endControlFlow()
            .addStatement("return value.${element.kotlinAccessor}.matches(Regex($fieldName))")
            .build()
    }
}
